logging/net_appender.go

```diff
@@ -206,40 +207,45 @@ func (nl *NetAppender) backgroundWorker() {
 	}
 }
 
-// Sync will flush the internal buffer of logs.
-func (nl *NetAppender) Sync() error {
-	for {
-		batch := func() []*apppb.LogEntry {
-			nl.toLogMutex.Lock()
-			defer nl.toLogMutex.Unlock()
-
-			if len(nl.toLog) == 0 {
-				return nil
-			}
-
-			batchSize := writeBatchSize
-			if len(nl.toLog) < writeBatchSize {
-				batchSize = len(nl.toLog)
-			}
+// Returns whether there is more work to do or if an error was encountered while trying to ship logs
+// over the network.
+func (nl *NetAppender) syncOnce() (bool, error) {
+	nl.toLogMutex.Lock()
+	defer nl.toLogMutex.Unlock()
 
-			ret := nl.toLog[:batchSize]
-			nl.toLog = nl.toLog[batchSize:]
+	if len(nl.toLog) == 0 {
+		return false, nil
+	}
 
-			return ret
-		}()
+	batchSize := writeBatchSize
+	if len(nl.toLog) < writeBatchSize {
+		batchSize = len(nl.toLog)
+	}
 
-		if len(batch) == 0 {
-			return nil
-		}
+	batch := nl.toLog[:batchSize]
+	err := nl.remoteWriter.write(batch)
+	if err == nil {
+		// On success remove the batch from the queue.
+		nl.toLog = nl.toLog[batchSize:]
```

Comment:
Because we hold the lock the entire time now, I only pop the batch after success. This removes the (real) out of order bug.